Sentry error reporting

chromakode opened the pull request:
Here's a roughly complete Sentry setup for both the client and server. Errors include the git commit and the user's id (on the client side). Once source maps are up and public, Sentry will be able to display full tracebacks for minified client errors.

I needed to change the scope of the server docker volume so that I could pipe git commit hashes through to the backend -- there are other ways to do this, but it seemed like the more straightforward approach.

A few caveats:
 * This may not cover all errors -- it should pick up global errors and errors propagated to express middleware, but there might be some areas of the code that will need extra instrumentation.
 * While testing the backend logging, I have seen Sentry fail to scrub a password in a register request body. I believe it's set up properly, and this code seems bog-standard, so I've asked them for clarification about this.

chromakode committed:
Change root of server volume to include full git repo

This is so we're able to access the git commit from the server execution
environment.

chromakode committed:
Send server-side errors to Sentry

chromakode committed:
Send client-side errors to Sentry

chromakode committed:
Output source maps where Sentry can find them

chromakode commented on client/scripts/constants.js:
```diff
@@ -41,5 +41,6 @@ export default {
   FIREBASE_WEB_KEY: 'AIzaSyCkPlC2qRkXchd9AdubS6aAyvhE1TNAPqU',
-  FIREBASE_SENDER_ID: '820872076821'
+  FIREBASE_SENDER_ID: '820872076821',
+  SENTRY_DSN: 'https://3ef3bbac14984711984c884138df8474@sentry.io/195290'
```
This will need to be changed to reflect your Sentry account.

davidgljay replied:
Got it!

chromakode commented on server/src/app.js:
```diff
@@ -20,12 +21,22 @@ const passport = require('passport')
 const startSubscriptionServer = require('./graph/subscriptions/SubscriptionServer')
 const PORT = 8181
 
+const GIT_HASH = execSync('git rev-parse HEAD', {
```
Running a synchronous shell command at server spinup feels icky to me. Not sure what your prod setup looks like, but a better approach might be to build a docker image for every backend deploy and embed the git commit during the image build step. Could also include the docker image hash in that case (possibly w/ Sentry's release tracking functionality).

davidgljay replied:
Agreed, there are a few initialization steps like this that need to be moved to docker image creation, it shouldn't be hard to move over but it hasn't been a priority.

chromakode committed:
Cleanup

chromakode commented:
Sentry also has a nifty user feedback form you might want to set up: https://blog.sentry.io/2016/04/21/introducing-user-feedback.html

chromakode committed:
Embed raven sentry reporter in main bundle

davidgljay commented on client/scripts/app.js:
```diff
@@ -57,6 +66,13 @@ class Nametag extends Component {
     }
+
+    // Attach user id to Sentry error reports
+    client.watchQuery({query: USER_QUERY}).subscribe({
```
Is it alright if I move this to a Redux function? In all cases where the user is logged in we're already getting this data, and I'd like to avoid having a duplicative call to the server. Creating a function that's something like ``setSentryUser(data.me.id)`` that can be called when that data is available might be cleaner.

Happy to pair to show you where this happens if you'd like to do it.

chromakode replied:
Sure -- I don't have a full view of the codebase so I anticipated this might come up in review. I looked around the code store but didn't spot a good place to put it there.

Does [`watchQuery`](http://dev.apollodata.com/core/apollo-client-api.html#ApolloClient.watchQuery) kick off a fetch to a query? The Apollo docs were a bit hard to understand, but I thought that it was akin to setting a watch on the data being fetched when it was triggered via other means.

If you'd like, feel free to put a commit on this branch moving this to where you'd like, or if you can point me in the right direction I'll do it.

davidgljay replied:
Ahh, looks like setting a fetchPolicy of `fetchOnly` will do the trick. I'll submit a commit.

davidgljay committed:
Adding fetch policy to watchQuery

This should prevent Raven from creating a duplicative call to the server to retrieve user data.

davidgljay commented:
Merging PR! Thanks @chromakode

davidgljay committed:
Merge branch 'master' into sentry
